Initial implementation

maximopalopoli opened the pull request:
An initial implementation, with the game logic and a little interaction with the user.

maximopalopoli committed:
Create a grid and pass an initial seed

maximopalopoli committed:
Added clock and print logic

maximopalopoli committed:
Add tests and implement start function

maximopalopoli committed:
Fix clippy issues and add a more complex test

maximopalopoli committed:
Name tests and change use in mod tests

maximopalopoli committed:
Refactor - create grid.rs file for grid type

maximopalopoli committed:
cargo fmt

maximopalopoli committed:
Make user interaction with command start

JereSalo commented on README.md:
```diff
@@ -1,3 +1,11 @@
+
+To run the game, use:
+
+`cargo run start 1y 1x 2y 2x ... Ny Nx`
```
```suggestion
`cargo run start 1x 1y 2x 2y ... Ny Nx`
```
In your code x goes first
Also, I think that a little explanation of parameters' meaning would be useful for newcomers, just say that those are the coordinates of alive cells.

maximopalopoli replied:
Yes, I had a confusion about this because the traditional x-y axes are rotated in the drawing. For the UI version I will clarify which is X and which is Y.

JereSalo commented on src/grid.rs:
```diff
@@ -0,0 +1,269 @@
+
+impl Grid {
+    pub fn new(width: usize, height: usize) -> Self {
+        let mut cells: Vec<Vec<u8>> = Vec::with_capacity(width);
+        for _ in 0..width {
+            cells.push(vec![0; height]);
+        }
```
```suggestion
        let cells: Vec<Vec<u8>> = vec![vec![0; height]; width];
```
I think this looks better.

JereSalo commented on src/grid.rs:
```diff
@@ -0,0 +1,269 @@
+use std::time::Duration;
+
+pub struct Grid {
+    matrix: Vec<Vec<u8>>,
```
This works but maybe it would be a better alternative to use other type of value for representing cell status like a boolean or an enum with 2 possible values. Using u8 for representing only 2 states may be a little bit overkill, but that's just my POV

maximopalopoli replied:
Yes, I preferred an u8 because initially because to print the values on the grid it was easier to print an 1 or 0 than a true or false, but now its better a boolean value.

JereSalo commented on src/grid.rs:
```diff
@@ -0,0 +1,269 @@
+    use super::Grid;
+
+    #[test]
+    fn test_01_new_returns_a_valid_matrix() {
```
test_ prefix is not necessary here. You can remove it if you want to.

JereSalo commented on src/grid.rs:
```diff
@@ -0,0 +1,269 @@
+
+        for (x, row) in new_matrix.iter_mut().enumerate().take(self.width) {
+            for (y, field) in row.iter_mut().enumerate().take(self.width) {
+                let neighbours_ammount = self.count_neighbours(x, y);
```
```suggestion
                let neighbours_amount = self.count_neighbours(x, y);
```
The word amount contains only one 'm'

JereSalo commented on src/grid.rs:
```diff
@@ -0,0 +1,269 @@
+        }
+    }
+
+    fn count_neighbours(&self, x: usize, y: usize) -> u8 {
```
This function is 66 lines long. Do you think it can be shortened in any way?

JereSalo commented on README.md:
```diff
@@ -1,3 +1,11 @@
+
+The start functionality will print the status of the current generation, and advance to the next generation every 3 seconds.
+
+In the next versions, there will be easier to pick the points, via UI.
```
```suggestion
In the next versions, it will be easier to select the points, via UI.
```

JereSalo commented on README.md:
```diff
@@ -1,3 +1,11 @@
 # Conways-Game-Of-Life
 
 An implemention in Rust of the [Game of Life](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life) made by John Conway. This is work in progress.
```
```suggestion
An implementation in Rust of the [Game of Life](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life) made by John Conway. This is work in progress.
```

JereSalo commented on src/grid.rs:
```diff
@@ -0,0 +1,269 @@
+    }
+
+    #[test]
+    fn test_04_clock_doesnt_make_die_cells_without_underpopulation_or_overpopulation() {
```
```suggestion
    fn clock_doesnt_kill_cells_without_underpopulation_or_overpopulation() {
```
This test (along with some previous ones) could have a better name. The phrase ‘Doesn’t make die cells’ sounds awkward. I’ve provided a suggestion, but feel free to rename it as you see fit.

JereSalo commented on src/main.rs:
```diff
@@ -0,0 +1,30 @@
+        .chunks(2)
+        .map(|pair| {
+            let x: usize = pair[0].parse().expect(
+                "Error parseando el primer valor de la coordenada",
```
```suggestion
                "Error parsing first coordinate value",
```

JereSalo reviewed (changes requested):
I don't think that it was necessary in this PR to have the print on screen functionality nor the `cargo run start`. With just the tests and basic structure and functionality of the grid it was good enough. At the end of the day you are going to replace that `print()` with something else, so even though it may be nice to see it on screen it isn't fundamental. Of course it adds value but I would have only focused on the basic functionalities and not the UI.
I like the variety of tests that you included and the grid structure is well done.
I think you've done a great job, there are just a few changes to make, but it looks pretty good.

maximopalopoli committed:
Change issues after PR review

The main changes are:
- Some little changes in the Readme.md
- The value that stores the life of the cell changes from u8 to boolean
- The logic of count_neighbours has changed
- The bounds error is absorbed by the limited_range_for_number function

maximopalopoli committed:
Clippy + fmt

JereSalo reviewed (approved):
It looks better now. I ran `cargo fmt` and it formatted a little thing but everything seems fine! Maybe I would add some small comments in functions to say what they do just for the reviewer to be able to understand the code faster.
